File: javascript/src/binary-search.js
// recursive style
function chopRecursion(target, array) {

	var midPoint = parseInt(array.length/2);

    if(array.length == 0 || array.length == 1 && target != array[midPoint])
    {
        return -1;
    }
	else if(target > array[midPoint])
    {
        return getResult(midPoint, chopRecursion(target, array.slice(midPoint, array.length)));
	}
	else if(target < array[midPoint])
    {
        return getResult(0, chopRecursion(target, array.slice(0, midPoint)));
	}
	else
    {
		return midPoint;
	}
}

function getResult(mid, position){
    return position == -1 ? -1 : mid + position;
}

// iteration style
function chopIteration(target, array){
    var left = 0;
    var mid = parseInt(array.length/2);
    var right = array.length;

    while(left < right){
        if(target < array[mid])
        {
            right = mid;
        }
        else if(target > array[mid])
        {
            left = mid + 1;
        }
        else
        {
            return mid;
        }
        mid = parseInt((left + right)/2);
    }
    return -1;
}

// function call style
function chopFunctionCall(target, array){
    return binarySearch(target, array, 0, array.length);
}

function binarySearch(target, array, left, right){
    var mid = parseInt((left + right)/2);

    if(left < right)
    {
        if(target > array[mid])
        {
            return binarySearch(target, array, mid+1, right);
        }
        else if(target < array[mid])
        {
            return binarySearch(target, array, left, mid);
        }
        else
        {
            return mid;
        }
    }
    else
    {
        return -1;
    }
}

File: csharp/kata/src/BinarySearch.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace kata
{
    public class BinarySearch
    {
        public int ChopRecursion(int target, int[] array)
        {
            try
            {
                return ChopWithException(target, array);
            }
            catch(Exception)
            {
                return -1;
            }

        }

        private static int ChopWithException(int target, int[] array)
        {
            var mid = array.Length/2;

            if (array.Length == 0)
            {
                throw new Exception();
            }
            if (target > array[mid])
            {
                return mid + ChopWithException(target, GetRightHalf(mid, array)) + 1;
            }
            if (target < array[mid])
            {
                return ChopWithException(target, GetLeftHalf(mid, array));
            }
            return mid;
        }

        private static int[] GetLeftHalf(int mid, int[] array)
        {
            return array.Take(mid).ToArray();
        }

        private static int[] GetRightHalf(int mid, int[] array)
        {
            return array.Skip(mid + 1).ToArray();
        }

        public int ChopIteration(int target, int[] array)
        {
            int left = 0;
            int mid = array.Length/2;
            int right = array.Length;

            while(left < right){
                if(target < array[mid])
                {
                    right = mid;
                }
                else if(target > array[mid])
                {
                    left = mid + 1;
                }
                else
                {
                    return mid;
                }
                mid = (left + right)/2;
            }
            return -1;
        }
    }
}
